Fix redundant ilp calls

srini229 opened the pull request.

854768750 commented on PlaceRouteHierFlow/placer/Placer.cpp:
```diff
@@ -241,30 +241,47 @@ std::map<double, std::pair<SeqPair, ILP_solver>> Placer::PlacementCoreAspectRati
+      ILP_solver tsol(designData, hyper.ilp_solver);
+      curr_cost = tsol.GenerateValidSolution(designData, curr_sp, drcInfo, hyper.NUM_THREADS);
+      if (curr_cost >= 0) {
+        oData[curr_cost] = std::make_pair(curr_sp, tsol);
+        ReshapeSeqPairMap(oData, nodeSize);
+      }
+      exhausted = curr_sp.EnumExhausted();
+      curr_sp.PerturbationNew(designData);
+      if (cnt >= maxcount) break;
+    }
+    if (curr_sp.EnumExhausted()) {
+      logger->info("Exhausted all permutations of sequence pairs");
+      return oData;
+    }
+  }
+
   while (++trial_count < hyper.max_init_trial_count) {
     // curr_cost negative means infeasible (do not satisfy placement constraints)
     // Only positive curr_cost value is accepted.
-    if (hyper.select_in_ILP)
-      curr_cost = curr_sol.GenerateValidSolution_select(designData, curr_sp, drcInfo);
-    else
-      curr_cost = curr_sol.GenerateValidSolution(designData, curr_sp, drcInfo, hyper.NUM_THREADS);
 
-    curr_sp.cacheSeq(designData);
+    if (!curr_sp.isSeqInCache(designData)) {
+      if (hyper.select_in_ILP)
+        curr_cost = curr_sol.GenerateValidSolution_select(designData, curr_sp, drcInfo);
+      else
+        curr_cost = curr_sol.GenerateValidSolution(designData, curr_sp, drcInfo, hyper.NUM_THREADS);
+      curr_sol.cost = curr_cost;
+      curr_sp.cacheSeq(designData);
+    }
```
looks good

srini229 commented:
There might be a failure in CI for the test: `tests/test_cmdline.py::test_cmdline[cascode_current_mirror_ota-2-False-False]`. This appears to be a router issue. There is an M3 shape created by the router that off-grid.

srini229 commented:
@854768750 There is a failing test in this branch most likely caused by the router. The placement checker did not complain. So I presume all placement constraints are satisfied. Can you take a look at the routing issue? The router seems to generate an off-grid M3 shape that the checker detects as "DIFFERENT WIDTH".
You can reproduce the issue using:

`pytest -s -v "tests/test_cmdline.py::test_cmdline[cascode_current_mirror_ota-2-False-False]"`

or:

`schematic2layout.py $ALIGN_HOME/examples/cascode_current_mirror_ota -p $ALIGN_HOME/pdks/FinFET14nm_Mock_PDK -flat 0 -v INFO -l INFO -n 2`

854768750 commented:
Yes I will take a look

854768750 commented:
The two blocks: X_M1NUP and X_M1PDOWN, are not place on grid, whose LL are (256, 13608) and (1024, 13608), respectively.
M1 pitch is 80

@soneryaldiz I am also wondering why checker did not report placement offgrid. 
@srini229 I can look into why placement offgrid.

soneryaldiz commented:
@srini229 , should we merge master to this branch? Some of the file changes do not look right.

854768750 commented:
> @854768750 , I am not sure if this PR has the latest offset fixing code.. @srini229 , should we merge master to this branch? Some of the file changes do not look right.

Ok this may a possible reason

854768750 commented:
@srini229 please check this code change https://github.com/ALIGN-analoglayout/ALIGN-public/pull/1180/commits/80b4604ad42e713dca42715ce29d9d1565c8275a.
After I remove this change, placement is on grid.

@soneryaldiz Is it possible to find why the checker did not report placement offgrid?

srini229 commented on PlaceRouteHierFlow/placer/ILP_solver.cpp:
```diff
@@ -2795,10 +2915,15 @@ double ILP_solver::GenerateValidSolution(const design& mydesign, const SeqPair&
     }
   }
 
+  if (snapGridILP && !GenerateValidSolutionCore(mydesign, curr_sp, drcInfo, num_threads, true)) return -1;
```
@854768750 This call here repeats ILP if any block is off-grid and snaps them back to grid. We don't want to go into placeongrid mode always [`here`](https://github.com/ALIGN-analoglayout/ALIGN-public/blob/80b4604ad42e713dca42715ce29d9d1565c8275a/PlaceRouteHierFlow/placer/design.cpp#L79-L84). That is costing too much in runtime.

854768750 replied:
ok, my understanding is that:
1. If no block has offset, then no need to snapGridILP. In this case, rounding after compaction is required.
2. If any block has offset, snapGridILP is true. Then no need to do rounding

854768750 replied:
In this testcase, no block has offset, then snapGridILP is false, compaction does not consider on grid, and rounding was removed, so some block are off grid.
Some options:
1. Always add offset==0 for empty offset, to activate on grid compaction
2. resume rounding for empty offset

srini229 replied:
> ok, my understanding is that:
> 
>     1. If no block has offset, then no need to snapGridILP. In this case, rounding after compaction is required.
> 
>     2. If any block has offset, snapGridILP is true. Then no need to do rounding

I removed compaction totally.
If there is no offset constraint and if any block is off-grid, we call ILP a second time setting snap to grid in ILP mode.
If any offset is present for any block, we just ever to the zero offset constraint that you had added in the previous PR.

854768750 commented on PlaceRouteHierFlow/placer/design.cpp:
```diff
@@ -41,6 +41,18 @@ design::design(PnRDB::hierNode& node, PnRDB::Drc_info& drcInfo, const int seed)
+  bool offsetpresent{false};
+  for (vector<PnRDB::blockComplex>::iterator it = node.Blocks.begin(); it != node.Blocks.end(); ++it) {
+    for (int bb = 0; bb < it->instNum; ++bb) {
+      if ((it->instance).at(bb).xoffset.size() > 0
```
Why only xoffset ? No yoffset?

srini229 replied:
you are right. needs both.

srini229 commented:
@soneryaldiz can you try if the latest check-in helps with the failing test with ILP infeasibility?

srini229 commented:
SA placer is unable to find a solution in 100 iterations in the failing integration test:
`tests/integration/test_integration.py::test_integration[1-top_down-True-100-1-FinFET14nm_Mock_PDK-switched_capacitor_filter]`
 I'll try to debug. To proceed, we can increase the number of iterations to 1000 for this test alone.

soneryaldiz commented:
@srini229
- The smaller test case works now. The ILP cost makes sense. The enumeration still keeps shuffling the auxiliary symmetry line variable. This results in 400 enumarations instead of just 4. Is there a way to avoid this auxiliary variable or detect when it is redundant?
- The larger test case still fails for the same subcircuit. The log still shows cost being nan for various subblocks.

srini229 commented:
@soneryaldiz I'm working on getting rid of the auxiliary variable for symmetry at least for the enumeration mode.
Regarding the NaN, can we try to debug again using the env variable `ALIGN_DEBUG_SA_ILP`? As we saw, it is most likely due to a problematic ILP constraint.

soneryaldiz commented:
@srini229 , just to clarify, I did not mean to ignore all constraints from enumeration but just the redundant ones. The issue should be in PlaceOnGrid constraints as the layout is generated fine when I disable it. To make faster progress, I will try to reproduce this on mock PDK.

srini229 commented:
@soneryaldiz ordering is taken care of in enumeration mode using the topo-sort. Symmetry and alignment constraints are anyway enforced in ILP. So we do not need to check for any of these constraints in enum mode.

For the failing test, is there any dummy layer in the layers.json with a negative pitch?

soneryaldiz commented:
@srini229 , there was a bug on the `new_example_circuit` please pull and merge again.

srini229 commented:
@soneryaldiz the test passes now and the PR is ready for review. @854768750

soneryaldiz commented on PlaceRouteHierFlow/placer/design.cpp:
```diff
@@ -41,12 +41,29 @@ design::design(PnRDB::hierNode& node, PnRDB::Drc_info& drcInfo, const int seed)
+  }
+  int gridx_pitch = drcInfo.Metal_info[v_metal_index].grid_unit_x;
+  int gridy_pitch = drcInfo.Metal_info[h_metal_index].grid_unit_y;
+  logger->info("grid pitches : {0} {1}", gridx_pitch, gridy_pitch);
```
this should be debug

854768750 commented on PlaceRouteHierFlow/placer/ILP_solver.cpp:
```diff
@@ -1098,7 +1099,7 @@ double ILP_solver::GenerateValidSolutionAnalytical(design& mydesign, PnRDB::Drc_
         if (mydesign.Blocks[iter2][0].blockPins.size() > 0) {
           for (const auto& centerk : mydesign.Blocks[iter2][0].blockPins[iter].center) {
             // calculate contact center
-            int pin_x = centerk.x;
+            int  pin_x = centerk.x;
```
an extra space

soneryaldiz commented on tests/gdsconv/black_box_test/test_black_box.py:
```diff
@@ -19,7 +19,7 @@
     ALIGN_WORK_DIR = pathlib.Path(os.environ['ALIGN_WORK_DIR']).resolve()
 
-@pytest.mark.nightly
+@pytest.mark.skip
```
Why skip this test? Please add a reason or bring it back.

soneryaldiz replied:
Sounds good. Could you please just annotate reason to skip then?

srini229 replied:
Please check the latest commit that resolves all the reviews.

soneryaldiz commented on tests/gdsconv/black_box_test/hsc_black_box_test/high_speed_comparator.const.json:
```diff
@@ -1,33 +1,140 @@
-    {"constraint": "Align",
-        "line" : "h_bottom",
-        "instances": ["xi0", "ccp", "xi1"]
+    {
```
What is the reason for changing this file?

srini229 replied:
changed it to keep it same as the one in examples/ directory.

soneryaldiz commented on align/pnr/placer.py:
```diff
@@ -39,6 +39,7 @@ def place( *, DB, opath, fpath, numLayout, effort, idx, lambda_coeff, select_in_
     hyper.LAMBDA = lambda_coeff
     hyper.use_analytical_placer = use_analytical_placer
     hyper.use_ILP_placer = place_using_ILP
+    hyper.ILP_runtime_limit = placer_ilp_runtime
```
Could you please add comment to explain what this parameter does, what the unit is, etc.?

soneryaldiz commented on PlaceRouteHierFlow/placer/ILP_solver.cpp:
```diff
@@ -2969,13 +3094,14 @@ double ILP_solver::GenerateValidSolution(const design& mydesign, const SeqPair&
-    // logger->debug("ILP__HPWL_compare : HPWL_extend={0} HPWL_ILP={1}", HPWL_extend, HPWL_ILP);
-    // logger->debug("ILP__Area_compare : area={0} area_ilp={1}", area, area_ilp);
-  }
+  //if (cost >= 0.) {
```
Should we assert that cost is not NaN?

soneryaldiz commented on PlaceRouteHierFlow/placer/Placer.cpp:
```diff
@@ -241,6 +242,28 @@ std::map<double, std::pair<SeqPair, ILP_solver>> Placer::PlacementCoreAspectRati
+      }
+      curr_sp.PerturbationNew(designData);
+      ++cnt;
+      if (cnt >= maxcount) break;
```
1. After `maxcount` trials, do we want to fall back to Simulated Annealing for another `maxcount` trials? Should we just throw an assertion instead? 
2. If we still want to resort to Simulated Annealing, should we cache these enumerations so that we do not solve ILP again?

soneryaldiz replied:
The cost comparisons in this method do not seem consistent, both >=0 and >0 are used.

srini229 replied:
I think `assert()` makes sense here. Since there aren't any more sequence pairs left to explore.

soneryaldiz reviewed (approved):
@srini229 , thank you for doing this. It would be icing on the cake if you could add a short description of tasks completed in this PR.

srini229 commented:
List of tasks completed in this PR:
- Added a runtime limit option for the ILP in placer (time in seconds)
- When no `PlaceOnGrid` constraints are present, by default do not snap block coordinates to `M1/M2` pitch grids. If the solution found violates the pitch grids, then call ILP again with grid snapping constraints added.
These constraints add: $$x=n_1 \times pitch_x,~~ y=n_2 \times pitch_y$$
- When `PlaceOnGrid` constraints are present, snapping blocks to pitch grids is done by adding a zero offset `PlaceOnGrid` constraint for blocks with no constraint.
- `PlaceOnGrid` parser in the placer only read the first constraint and ignored the rest for any hierarchy. Fixed it by reading in all the constraints.
- Enum mode is separated from the SA placer; Iterate over all legal seq pairs in enum mode and return the least cost solutions.
- Ordering is implicitly handled in enum mode using the ordered enumerator; symmetry/keep ordering routines are not required and are hence skipped.
- Seq pair caching now caches cost as well. The placement information is not cached considering the potential explosion in memory to cache the coordinates of all blocks.
- Added support for the `lp_solve` solver for runtime comparison.
- Fixed the ordered enumerator to consider selection variants.
- In enum mode, auxiliary symmetry variables are not required and hence are no longer added to the seq pair.
